MobiusCore/Source/MobiusLoop.swift

```diff
@@ -19,39 +19,77 @@
-    private let disposable: Disposable
-    private var disposed = false
-    private var access: ConcurrentAccessDetector
+/// A `MobiusLoop` is the core encapsulation of business logic in Mobius.
+///
+/// It stores a current model, applies incoming events, processes the resulting effects, and broadcasts model changes
+/// to observers.
+///
+/// Use `Mobius.loop(update:effectHandler:)` to create an instance.
+public final class MobiusLoop<Model, Event, Effect>: Disposable {
+    private var access = ConcurrentAccessDetector()
     private var workBag: WorkBag
 
-    public var debugDescription: String {
-        return access.guard {
-            if disposed {
-                return "disposed \(type(of: self))!"
-            }
-            return "\(type(of: self)) \(eventProcessor)"
-        }
-    }
+    private var effectConnection: Connection<Effect>! = nil
+    private var consumeEvent: Consumer<Event>! = nil
+    private let modelPublisher: ConnectablePublisher<Model>
+
+    private var model: Model
+
+    private var disposable: CompositeDisposable?
 
     init(
-        eventProcessor: EventProcessor<Model, Event, Effect>,
-        consumeEvent: @escaping Consumer<Event>,
-        modelPublisher: ConnectablePublisher<Model>,
-        disposable: Disposable,
-        accessGuard: ConcurrentAccessDetector,
-        workBag: WorkBag
+        model: Model,
+        update: Update<Model, Event, Effect>,
+        eventSource: AnyEventSource<Event>,
+        eventConsumerTransformer: ConsumerTransformer<Event>,
+        effectHandler: AnyConnectable<Effect, Event>,
+        effects: [Effect],
+        logger: AnyMobiusLogger<Model, Event, Effect>
     ) {
-        self.eventProcessor = eventProcessor
-        self.consumeEvent = consumeEvent
-        self.modelPublisher = modelPublisher
-        self.disposable = disposable
-        self.access = accessGuard
+        let loggingUpdate = update.logging(logger)
+
+        let workBag = WorkBag(accessGuard: access)
         self.workBag = workBag
+
+        self.model = model
+        self.modelPublisher = ConnectablePublisher<Model>()
+
+        // consumeEvent is the closure that processes an event and handles the model and effect updates. It needs to
+        // be a closure so that it can be transformed by eventConsumerTransformer, and to handle ownership correctly:
+        // consumeEvent holds on to the update function and workbag, and also holds self while its work bag submission
+        // is queued.
+        //
+        // Originally the processNext(...) invocation was wrapped in a method, but that just spread things out more.
+        let consumeEvent = eventConsumerTransformer { [unowned self] event in
+            // Note: captures self strongly until the block is serviced by the workBag
+            let processNext = self.processNext
+            workBag.submit {
+                processNext(loggingUpdate.update(model: self.model, event: event))
```

Comment:
This is hard for me to parse - but I hope it captures the value of `self.model` at actual execution time, not at 'event consume' time. That is, it must be the case that event E2 is processed with the model resulting from processing event E1, whether E2 is submitted before or after E1 has finished processing. Is that correct?

Reply:
Yes, this is correct. If we had `let model = self.model` outside the `workBag.submit` and used it here, we would hit the issue you’re describing (and I can verify that existing unit tests fail in that scenario). As written, `self.model` is read immediately before calling `processNext` (this is an unowned read of `self`, but the lifetime is guaranteed by the curried `processNext`).

Reply:
I’ll add a comment to clarify this

Reply:
Comment added BTW

Reply:
Looks good!